### PR TITLE
fix(core): divide by true mass and fall back to uniform in _normalize_simplex (#2470)

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2470.

## Summary
In `alberta_framework/core/upgd_memory.py`, `_normalize_simplex` divided non-negative mass by `jnp.maximum(jnp.sum(clipped), 1e-12)`. When total mass was zero, negative, or positive below the `1e-12` floor (such as when `state.upgd_state.previous_targets` is all zeros at initialization), the returned array did not sum to 1.0 (reporting 0.0 or a shrunken mass).

## Proposed Changes
1. Updated `_normalize_simplex` to divide by the true finite positive sum `total`, and fall back to the uniform distribution `1.0 / clipped.shape[0]` when `total <= 0.0` or non-finite.
2. Added unit tests in `tests/test_upgd_memory.py` asserting that `_normalize_simplex` sums to 1.0 across zero vectors, negative vectors, and sub-1e-12 small positive vectors.

## Test Plan
- `uv run pytest tests/test_upgd_memory.py` (137 passed)
- `uv run ruff check alberta_framework/core/upgd_memory.py tests/test_upgd_memory.py` (clean)
- `uv run mypy alberta_framework/core/upgd_memory.py` (clean)

### CODE REVIEW

**Review by agent-cortex — APPROVED** [review]
**Approve.** I independently reproduced the defect and the fix, and verified the new test discriminates. Reviewed at head `68787a2` against base `c9aba7b` (28→ `_normalize_simplex` in `alberta_framework/core/upgd_memory.py`).

## Scope
Tight and on-target for #2470: one function body (`upgd_memory.py:676`) plus one parametrized test (`tests/test_upgd_memory.py`). No manifest/lockfile/CI/hook changes. All three live callers pass 1-D arrays (`upgd_memory.py:968,984,987`), so the function's 1-D contract is respected.

## The defect is real (base fails)
`_normalize_simplex` is consumed in `softmax_ce` readout blending, where its output is expected to be a valid probability simplex. On base, `clipped / jnp.maximum(jnp.sum(clipped), 1e-12)` does **not** normalize degenerate or tiny-mass inputs. Running the base body verbatim under jax:

```
case                           |   base sum |   head sum
all-zero      [0,0,0]          |   0.000000 |   1.000000
all-negative  [-1,-2,-3]       |   0.000000 |   1.000000
tiny-below1e12[7.5e-13,0,0]    |   0.750000 |   1.000000
tinier        [1e-30,0,0]      |   0.000000 |   1.000000
normal        [1,1,2]          |   1.000000 |   1.000000
```

So on base a zero/negative prediction produces the all-zeros vector (sum 0), and a sub-`1e-12` mass gets divided by the `1e-12` floor rather than its true mass (sum 0.75) — both are invalid simplices fed into the downstream blend.

## The fix resolves it (head passes) without regressing the normal path
The exact assertion from the new test (`chex.assert_trees_all_close(sum, 1.0, atol=1e-6)`) on all four fixtures:

```
HEAD  all-zero      [0,0,0]        sum=1.0000000  test_pass=True
BASE  all-zero      [0,0,0]        sum=0.0000000  test_pass=False
HEAD  all-negative  [-1,-2,-3]     sum=1.0000000  test_pass=True
HEAD  tiny [7.5e-13,0,0]           sum=1.0000000  test_pass=True
HEAD  tinier [1e-30,0,0]           sum=1.0000000  test_pass=True
```

Normal input `[1,1,2] → [0.25,0.25,0.5]` (sum 1) is unchanged, so the valid-mass path is untouched. Dividing by the *true* `total` (not a floor) is what makes the sub-`1e-12` and even `1e-30` cases normalize exactly.

## Robustness probes (head)
```
nan     head=[0.333,0.333,0.333]  sum=1.0
posinf  head=[0.333,0.333,0.333]  sum=1.0
```
NaN/`inf` inputs fall through to `uniform` because `jnp.isfinite(total)` is false. The inner `jnp.where(valid, total, 1.0)` denominator is the right defensive move: JAX evaluates both branches, so replacing the denominator with `1.0` when invalid prevents a NaN/inf from the unselected `clipped/total` branch leaking through the outer `where`. Verified — no NaN leaks.

## Test hygiene
The test file already imports `chex`, `jnp`, and `pytest`, and this PR correctly adds `_normalize_simplex` to the import block (`tests/test_upgd_memory.py:15`), so the new cases actually reference the patched symbol and will run in the suite. (I could not execute the repo's own `pytest` in my sandbox: the image only has Python 3.11, and `alberta_framework/core/optimizers.py:423` uses PEP-695 `class Optimizer[...]` generics that require the repo's declared Python ≥3.12 — a toolchain gap on my side, not a PR issue. I therefore ran the base and head function bodies verbatim under `jax 0.10.2` with the new test's exact assertions, shown above.)

## Non-blocking notes (not required for merge)
1. `uniform = jnp.full_like(clipped, 1.0/clipped.shape[0])` hard-codes the 1-D axis. This is correct for all current callers, but if `_normalize_simplex` is ever reused on a batched `(B, K)` array the uniform fallback would use `B` instead of `K`. A `shape[-1]` would future-proof it; optional.
2. This PR supersedes the sibling #2768 for the same issue. #2783 is the technically stronger of the two: it divides by the true mass, whereas #2768 clamps the denominator at `1e-38` and — for an ultra-subnormal mass in `(0, 1e-38)` — returns a normalized vector whose entries sum to <1 while still passing its own `valid_mass` guard, i.e. an invalid simplex. Recommend merging #2783 and closing #2768.

Correctness, robustness, and test-discrimination all verified; no blocking gaps found.

---
AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/eliza@9b93ec4374878428a7a53b24202dfa15a8dbb55b:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [eliza-swarm]

**Review by kenjohnscreates — CHANGES_REQUESTED** [review]
Independent verification against main @ c9aba7b5 (isolated worktree, import provenance asserted). This branch still returns a non-simplex on the issue's dominant-entry class, so I have to request changes.

**Measured on this branch:** `[1e38, 1, 1]` → `[0, 0, 0]`, sum **0.0** (main: same failure); `[1e38, 1e-8, 0]` → zeros as well. All other probed classes (zero, all-negative, inf-total, tiny, sub-floor, subnormal) correctly return uniform or exact simplexes, and well-formed inputs are bit-exact vs main.

Mechanism: `valid = isfinite(total) & (total > 0)` inspects only the *total* — for `[1e38,1,1]` the total is finite and positive, so the guarded division runs, and XLA's `clipped * reciprocal(total)` flushes the subnormal reciprocal to zero elementwise. The result is all-zero but the guard already said "valid", so it is returned as-is. The validity test has to look at the *outcome* of the normalization (its sum, as your own #2749/#2503 do) or rescale before summing (as the frexp variants do), not only at the input mass.

Nine competing PRs on this issue already handle this class (measured comparison to be posted on #2470); converging on one of them seems better than iterating here.

Reviewed with claude-code / anthropic / claude-fable-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

**Comment by lalalune** [comment]
Closing as superseded by #2831 after comparing the implementation diffs. The retained fix uses power-of-two rescaling to preserve finite extreme probability ratios, uniform fallback for unusable mass, and a live target-trace blend regression. Several alternatives use a direct sum/division that loses finite extreme mass; the other rescaled implementations duplicate this fix with narrower regression coverage or unrelated formatting changes. On current main the retained tests reproduce 10 failures; with its implementation all 146 UPGD-memory tests pass. Required CI has been approved on the retained PR; no evidence artifacts are promoted.
